pkg/kubectl/cmd/rollout/rollout_status.go

```diff
@@ -169,42 +183,67 @@ func (o *RolloutStatusOptions) Run() error {
 		return err
 	}
 
-	// check if deployment's has finished the rollout
-	status, done, err := statusViewer.Status(info.Namespace, info.Name, o.Revision)
-	if err != nil {
-		return err
-	}
-	fmt.Fprintf(o.Out, "%s", status)
-	if done {
-		return nil
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", info.Name).String()
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return dynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return dynamicClient.Resource(info.Mapping.Resource).Namespace(info.Namespace).Watch(options)
+		},
 	}
 
-	shouldWatch := o.Watch
-	if !shouldWatch {
-		return nil
-	}
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: info.Namespace, Name: info.Name})
+		if err != nil {
+			return true, err
+		}
+		if !exists {
+			// We need to make sure we see the object in the cache before we start waiting for events
+			// or we would be waiting for the timeout if such object didn't exist.
+			return true, apierrors.NewNotFound(mapping.Resource.GroupResource(), info.Name)
+		}
 
-	// watch for changes to the deployment
-	w, err := r.Watch(rv)
-	if err != nil {
-		return err
+		return false, nil
 	}
 
 	// if the rollout isn't done yet, keep watching deployment status
-	intr := interrupt.New(nil, w.Stop)
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), o.Timeout)
+	intr := interrupt.New(nil, cancel)
 	return intr.Run(func() error {
-		_, err := watch.Until(0, w, func(e watch.Event) (bool, error) {
-			// print deployment's status
-			status, done, err := statusViewer.Status(info.Namespace, info.Name, o.Revision)
-			if err != nil {
-				return false, err
-			}
-			fmt.Fprintf(o.Out, "%s", status)
-			// Quit waiting if the rollout is done
-			if done {
-				return true, nil
+		_, err = watchtools.UntilWithInformer(ctx, lw, &unstructured.Unstructured{}, 0, preconditionFunc, func(e watch.Event) (bool, error) {
```

Comment:
I'm not sure I like using `WithInformer` in the signature.  That's a detail of the implementation.  `UntilWithResync` is more appropriate.

I'm also uncomfortable with clients getting an `Added` event, because an informer is *not* a watch list.  We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

Reply:
> I'm not sure I like using WithInformer in the signature. That's a detail of the implementation. UntilWithResync is more appropriate.

UntilWithResync it is.

> I'm also uncomfortable with clients getting an Added event, because an informer is not a watch stream. We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

Could you elaborate on that idea?

> We shouldn't be pretending that Added and Modified make sense with an informer loop, because they don't.

I guess same could be said about Watch from RV "". You get Added event, although the object already existed. Why should this differ with an informer?

Reply:
@smarterclayton rename is done but I am gonna need you to explain your point with Added/Modified event in more detail though